handlers/validator.go

```diff
@@ -1729,32 +1765,60 @@ func ValidatorHistory(w http.ResponseWriter, r *http.Request) {
 			})
 			continue
 		}
-		events := template.HTML("")
-		if incomeDetails[index][i].AttestationSourcePenalty > 0 && incomeDetails[index][i].AttestationTargetPenalty > 0 {
-			events += utils.FormatAttestationStatusShort(2)
-		} else {
-			events += utils.FormatAttestationStatusShort(1)
+		event := types.ValidatorHistoryEvent{}
+
+		if incomeDetails[index][i].AttestationSourceReward > 0 {
+			event.AttestationSourceStatus = 1
+		} else if incomeDetails[index][i].AttestationSourcePenalty > 0 {
+			event.AttestationSourceStatus = 2
+		}
+
+		if incomeDetails[index][i].AttestationTargetReward > 0 {
+			event.AttestationTargetStatus = 1
+		} else if incomeDetails[index][i].AttestationTargetPenalty > 0 {
+			event.AttestationTargetStatus = 2
+		}
+
+		if incomeDetails[index][i].AttestationHeadReward > 0 {
+			event.AttestationHeadStatus = 1
 		}
 
 		if incomeDetails[index][i].ProposerAttestationInclusionReward > 0 {
-			block := utils.FormatBlockStatusShort(1)
-			events += block
+			event.BlockProposalStatus = 1
 		} else if incomeDetails[index][i].ProposalsMissed > 0 {
-			block := utils.FormatBlockStatusShort(2)
-			events += block
+			event.BlockProposalStatus = 2
+		}
+
+		if syncReward, syncPenalty := incomeDetails[index][i].SyncCommitteeReward, incomeDetails[index][i].SyncCommitteePenalty; syncReward > 0 {
+			if syncPenalty > 0 {
+				event.SyncParticipationStatus = 3
+				balanceChangeForSlot := (syncReward + syncPenalty) / utils.Config.Chain.Config.SlotsPerEpoch
```

Comment:
I did test this somewhat thoroughly, and while the sync rewards and penalties are not always the same, I've found that the floor of the ratio of penalty to reward times the slots per epoch will result in the correct amount of missed slots every time, but really that's just conjecture. The alternative is to fetch the correct amount of missed slots from SQL, but I feel like that could be too much of a performance downgrade. What are your thoughts?

Reply:
As discussed this calculation does not seem to work when there is a large percentage of missed syncs (e.g. 2/32).

I tested it on the prater network with validator 210684:

- Epoch 170305 shows 2/32 but should be 6/32
- Epoch 170309 shows 2/32 but should be 5/32

Reply:
I have now implemented it so that the number of proposed slots get fetched from the database if there are any sync rewards or penalties present in the set of epochs in the queried history page. With these, rewards and penalties are always the same.